Source/Core/Core/HW/SI/SI_DeviceGBA.cpp

```diff
@@ -233,140 +216,141 @@ void GBASockServer::ClockSync()
   }
 }
 
-void GBASockServer::Send(const u8* si_buffer)
+bool GBASockServer::Connect()
 {
-  if (!m_client)
-    if (!GetAvailableSock(m_client))
-      return;
+  if (!IsConnected())
+    m_client = GetNextSock();
+  return IsConnected();
+}
 
-  for (size_t i = 0; i < m_send_data.size(); i++)
-    m_send_data[i] = si_buffer[i ^ 3];
+bool GBASockServer::IsConnected()
+{
+  return static_cast<bool>(m_client);
+}
 
-  m_cmd = (u8)m_send_data[0];
+void GBASockServer::Send(const u8* si_buffer)
+{
+  if (!Connect())
+    return;
 
-#ifdef _DEBUG
-  NOTICE_LOG(SERIALINTERFACE, "%01d cmd %02x [> %02x%02x%02x%02x]", m_device_number,
-             (u8)m_send_data[0], (u8)m_send_data[1], (u8)m_send_data[2], (u8)m_send_data[3],
-             (u8)m_send_data[4]);
-#endif
+  std::array<u8, SEND_MAX_SIZE> send_data;
+  for (size_t i = 0; i < send_data.size(); i++)
+    send_data[i] = si_buffer[i ^ 3];
+
+  u8 cmd = send_data[0];
+  if (cmd != CMD_STATUS)
+    m_booted = true;
 
   m_client->setBlocking(false);
   sf::Socket::Status status;
-  if (m_cmd == CMD_WRITE)
-    status = m_client->send(m_send_data.data(), m_send_data.size());
+  if (cmd == CMD_WRITE)
+    status = m_client->send(send_data.data(), send_data.size());
   else
-    status = m_client->send(m_send_data.data(), 1);
-
-  if (m_cmd != CMD_STATUS)
-    m_booted = true;
+    status = m_client->send(send_data.data(), 1);
 
   if (status == sf::Socket::Disconnected)
     Disconnect();
-
-  m_time_cmd_sent = CoreTiming::GetTicks();
 }
 
 int GBASockServer::Receive(u8* si_buffer)
 {
   if (!m_client)
-    if (!GetAvailableSock(m_client))
-      return 5;
-
-  size_t num_received = 0;
-  u64 transfer_time = GetTransferTime((u8)m_send_data[0]);
-  bool block = (CoreTiming::GetTicks() - m_time_cmd_sent) > transfer_time;
-  if (m_cmd == CMD_STATUS && !m_booted)
-    block = false;
+    return 0;
 
-  if (block)
+  if (m_booted)
   {
     sf::SocketSelector selector;
     selector.add(*m_client);
     selector.wait(sf::milliseconds(1000));
   }
 
+  size_t num_received = 0;
+  std::array<u8, RECV_MAX_SIZE> recv_data;
   sf::Socket::Status recv_stat =
-      m_client->receive(m_recv_data.data(), m_recv_data.size(), num_received);
-  if (recv_stat == sf::Socket::Disconnected)
+      m_client->receive(recv_data.data(), recv_data.size(), num_received);
+  if (recv_stat == sf::Socket::Disconnected || recv_stat == sf::Socket::NotReady ||
```

Comment:
GBAs don't write back if they're not in JOYBUS mode. Disconnecting can cause multiple GBAs to swap positions if they're randomly disconnecting and reconnecting when they start being in JOYBUS mode.